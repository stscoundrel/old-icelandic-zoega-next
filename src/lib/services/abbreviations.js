import { findAbbreviations } from 'old-icelandic-zoega-abbreviations'
import { abbreviate } from 'abbreviatrix'

export const getAbbreviations = ({ definitions }) => {
  const combinedAbbreviations = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = findAbbreviations(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (content, abbreviations) => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
