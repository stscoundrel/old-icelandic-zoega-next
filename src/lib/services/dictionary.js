const { getDictionary } = require('old-icelandic-zoega')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let dictionaryCache = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (dictionaryCache) return dictionaryCache

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  dictionaryCache = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
