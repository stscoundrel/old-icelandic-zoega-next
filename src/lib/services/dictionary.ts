import { getDictionary, getNoMarkupDictionary } from 'old-icelandic-zoega'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import type { DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

// Light weight alternative when definitions are not needed.
// Next.js caches all props as JSON, so no need to store megabytes of definitions when not used.
export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let dictionaryCache: DictionaryEntry[] | null = null
let cachedInitialPages: string[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word)

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (dictionaryCache) return dictionaryCache

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  dictionaryCache = formattedWords

  return formattedWords
}

export const getAllWordsWithoutMarkup = (): DictionaryEntry[] => {
  const words = getNoMarkupDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  return formattedWords
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => ({ word: entry.word, slug: entry.slug }))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWordsWithoutMarkup()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

/**
 * Initial word pages to build are basically 6000
 * headword pages based on modulus. Larger number
 * can not be deployed in one go.
 */
export const getInitialWordsToBuild = (): string[] => {
  if (cachedInitialPages) return cachedInitialPages

  const allWords = getAllWords()

  const result: string[] = []
  for (let i = 0; i < allWords.length; i += 5) {
    result.push(allWords[i].slug);
  }

  cachedInitialPages = result
  return cachedInitialPages
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
