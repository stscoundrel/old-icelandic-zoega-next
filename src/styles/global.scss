@import 'reset-css';
@import 'variables';
@import 'media-queries';
@import 'typography';
@import 'containers';
@import 'buttons';

body,
html {
  max-width: 100%;
  overflow-x: hidden;
  scroll-behavior: smooth;
}

html {
  box-sizing: border-box;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}