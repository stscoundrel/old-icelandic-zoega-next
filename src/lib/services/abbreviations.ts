import { findAbbreviations } from 'old-icelandic-zoega-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from './dictionary'

export interface Abbreviation{
  abbreviation: string,
  explanation: string
}

export const getAbbreviations = ({ definitions }: DictionaryEntry): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = findAbbreviations(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations:
  Abbreviation[],
): string => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
