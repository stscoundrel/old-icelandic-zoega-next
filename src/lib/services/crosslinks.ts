import { getCrosslinks as getAllCrossLinks, Crosslink } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from './dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    // Only return foreign links, not ones pointing to self.
    return crosslinks[entry.slug].filter((link) => link.source !== 'old-icelandic')
  }

  return []
}

export default {
  getCrossLinks,
}
