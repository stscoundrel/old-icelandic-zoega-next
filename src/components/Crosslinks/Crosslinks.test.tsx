import renderer from 'react-test-renderer'
import { Crosslink } from 'lib/services/crosslinks'
import { DictionarySource } from 'scandinavian-dictionary-crosslinker'
import Crosslinks from './index'
import styles from './Crosslinks.module.scss'

describe('Crosslinks component', () => {
  const crosslinks: Crosslink[] = [
    {
      url: 'https://cleasby-vigfusson-dictionary.vercel.app/word/abbadis',
      source: DictionarySource.OldNorse,
    },
    {
      url: 'https://old-norwegian-dictionary.vercel.app/word/abbadis',
      source: DictionarySource.OldNorwegian,
    },
  ]

  test('Only renders when content is available', () => {
    const tree = renderer.create(
      <Crosslinks crosslinks={[]} />,
    )
    expect(tree.toJSON()).toBeNull()
  })

  test('Matches the snapshot', () => {
    const tree = renderer.create(
      <Crosslinks crosslinks={crosslinks} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct amount of crosslinks', () => {
    const tree = renderer.create(<Crosslinks crosslinks={crosslinks} />)
    const { root } = tree

    expect(root.findAllByProps({ className: styles.listItem }).length).toEqual(2)
  })

  test('Has expected crosslink content', () => {
    const tree = renderer.create(<Crosslinks crosslinks={crosslinks} />)

    expect(JSON.stringify(tree)).toContain('Old Norse - Cleasby & Vigfusson Dictionary')
    expect(JSON.stringify(tree)).toContain('Old Norwegian - Johan Fritzner\'s Dictionary')
  })
})
