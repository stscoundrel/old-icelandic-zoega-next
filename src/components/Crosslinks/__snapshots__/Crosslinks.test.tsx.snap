// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crosslinks component Matches the snapshot 1`] = `
[
  <h4>
    Also available in related dictionaries:
  </h4>,
  <p>
    This headword also appears in dictionaries of other languages closely related to Old Icelandic.
  </p>,
  <ul
    className="list"
  >
    <li
      className="listItem"
    >
      <a
        href="https://cleasby-vigfusson-dictionary.vercel.app/word/abbadis"
        rel="noopener noreferrer"
        target="_blank"
      >
        Old Norse - Cleasby & Vigfusson Dictionary
      </a>
    </li>
    <li
      className="listItem"
    >
      <a
        href="https://old-norwegian-dictionary.vercel.app/word/abbadis"
        rel="noopener noreferrer"
        target="_blank"
      >
        Old Norwegian - Johan Fritzner's Dictionary
      </a>
    </li>
  </ul>,
]
`;
